### PR TITLE
Add docker to yeoman generator

### PR DESCRIPTION
**Pull request by withinboredom**
This PR adds a simple `docker-compose.yml` file and a simple `Dockerfile` to the yeoman generator to allow people to quickly get up and running with Docker. This will be helpful for developers using Windows.

### CODE REVIEW

**Comment by gigabo** [comment]
Yeah! :metal: 

This is awesome!  I need to get set up with a modern docker to try it out.

@doug-wade - You're most familiar with this yeoman generator.  Take a look at _this_!

File: packages/generator-react-server/generators/app/templates/README.md [hunk 1]
```diff
@@ -4,15 +4,26 @@ Very simple example project for `react-server` which only shows server rendering
...
+```shell
 npm start
 ```
 
 Then go to [localhost:3000](http://localhost:3000/). You will see a simple page that pre-renders and that is interactive on load. It also will include hot reloading of React components in their own file.
 
 If you want to optimize the client code at the expense of startup time, type `NODE_ENV=production npm start`. You can also use any of [the other arguments for react-server-cli](../../react-server-cli#setting-options-manually) after `--`. For example:
 
-```
+```shell
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/README.md]: Thanks.  Nice to have syntax coloring on these.

File: packages/generator-react-server/generators/app/templates/README.md [hunk 2]
```diff
@@ -4,15 +4,26 @@ Very simple example project for `react-server` which only shows server rendering
...
 # start in dev mode on port 4000
 npm start -- --port=4000
 ```
+
+# Developing using Docker and Docker Compose
+
+These steps assume you are familiar with docker and already have it installed.
+
+Before getting started you'll need to configure the host to use your docker-machine or docker daemon. See [The docs](https://github.com/redfin/react-server/tree/master/packages/react-server-cli) for how to do this.
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/README.md]: Uh... maybe a _little_ more explanation for docker n00bs like me?  It's not clear to me what in the `react-server-cli` docs will help me with configuration for docker-machine or docker daemon. 😬

File: packages/generator-react-server/generators/app/templates/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+version: '2'
+services:
+  react:
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/docker-compose.yml]: Is this `react` just a label for the service?

If so, could we make it `react-server`, or `react_server` if dashes aren't allowed?

File: packages/generator-react-server/generators/app/templates/README.md (hunk 2 above)
Comment (withinboredom) [inline packages/generator-react-server/generators/app/templates/README.md]: Yeah, sure thing.

File: packages/generator-react-server/generators/app/templates/README.md [hunk 4]
```diff
@@ -4,15 +4,62 @@ Very simple example project for `react-server` which only shows server rendering
...
+2. Start `docker quick start shell`
+3. Navigate to where you generated the project
+4. Add a configuration to set the `host` option to the ip given by `docker-machine ip`. An example configuration might be like:
+```json
+{
+  "port": "3000",
+  "env": {
+    "docker": {
+      "host": "Your ip from `docker-machine ip` here"
+    },
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/README.md]: This is cool.  Nice use environment configuration.

Thanks for the added documentation.  Clears it right up for me.

File: packages/generator-react-server/generators/app/templates/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  react_server:                               # The label of the service
+    build: .                                  # The location of the Dockerfile to build
+    volumes:                                  # Files to share with the container and the host
+      - .:/www                                # Share the project in /www in the container
+      - react_server_node_modules:/www/node_modules  # A special volume so that OS specific node_modules are built correctly
+    working_dir: /www                         # The location all commands should run from
+    environment:
+      NODE_ENV: 'docker'                      # Set the NODE_ENV environment variable
+    command: /bin/bash -c "npm install && npm start" # The command to run in when the container starts
+    ports:                                    # Ports to expose to your host
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/docker-compose.yml]: All these comments are really helpful!

File: packages/generator-react-server/generators/app/templates/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+version: '2'
+services:
+  react_server:                               # The label of the service
+    build: .                                  # The location of the Dockerfile to build
+    volumes:                                  # Files to share with the container and the host
+      - .:/www                                # Share the project in /www in the container
+      - react_server_node_modules:/www/node_modules  # A special volume so that OS specific node_modules are built correctly
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/docker-compose.yml]: Oh, thanks for renaming the volume, too.  Didn't even catch that one.

**Comment by gigabo** [comment]
This looks great to me. 👍 

I'd like to wait until @doug-wade has had a chance to look over it, too, since he knows this generator well.

@doug-wade - Is this something that would be easy for us to prompt the user for down the road?  Like `want to run in a docker container? Yes/no` and conditionally include this stuff?

File: packages/generator-react-server/generators/app/templates/README.md [hunk 7]
```diff
@@ -4,15 +4,62 @@ Very simple example project for `react-server` which only shows server rendering
...
+    "staging": {
+      "port": "3000"
+    },
+    "production": {
+      "port": "80"
+    }
+  }
+}
```
Comment (gigabo) [inline packages/generator-react-server/generators/app/templates/README.md]: Just occurred to me that we should probably _generate_ this config.  https://github.com/redfin/react-server/issues/203

**Comment by gigabo** [comment]
Hey, I just tried this locally.  The README changes showed up, but the dockerfile and the docker-compose.yml didn't... @doug-wade is there something else that needs to happen for new files to get picked up?

**Comment by doug-wade** [comment]
@withinboredom This is pretty awesome!  As @gigabo notes, there are still some things we'll want you to do before you merge

1) I don't match files with globs currently, even [though I should](https://github.com/redfin/react-server/issues/204).  For now, you'll need to enumerate the new files in one of the [two lists of files](https://github.com/redfin/react-server/blob/master/packages/generator-react-server/generators/app/index.js#L26) that are operated on by the generator ([one](https://github.com/redfin/react-server/blob/master/packages/generator-react-server/generators/app/index.js#L26) is for dotfiles that are used by `git` and `babel` for configuration, which I preface with `_` rather than `.` to prevent them interacting with the monorepo; you may want to add your docker config there to make it easier to develop on Docker, though I'm not sure; if there isn't a risk of interaction, there is also [one](https://github.com/redfin/react-server/blob/master/packages/generator-react-server/generators/app/index.js#L38) for files that are not renamed.
2) Not all clients will want to include a Docker container with their new react-server project, so we should [prompt them](https://github.com/redfin/react-server/blob/master/packages/generator-react-server/generators/app/index.js#L12) (Yeoman uses [Inquirer for prompting](https://github.com/SBoudrias/Inquirer.js/) ), and only generate the Docker config if the user asks for it (my first intuition is to default to no Docker config, personally, since we push the Yeoman generator as an easy way to get started, and the Docker config makes react-server look bigger and scarier to a new developer; honestly [I should put the linter behind an Inquirer prompt](https://github.com/redfin/react-server/issues/205) for the same reason)

File: packages/generator-react-server/generators/app/templates/README.md (hunk 1 above)
Comment (doug-wade) [inline packages/generator-react-server/generators/app/templates/README.md]: 👍   Thanks for adding this!

**Comment by withinboredom** [comment]
@doug-wade,

I added the docker files as a prompt, and included changes to the unit tests. I can't get the new test (where `dockerCfg` is `true`) to pass, although I tested locally and it seems to function as expected. Maybe you can take a gander and give me a hint as to what I did wrong?

**Comment by withinboredom** [comment]
Also, it seems like a good idea to wait for #203 before generating the config for a dockerish and I think its a good idea to leave that in the README for now. What do you think?

**Comment by gigabo** [comment]
> wait for #203 before generating the config

Makes sense.

**Comment by gigabo** [comment]
Alright!  It's loading for me!

But... static assets aren't, yet.

image

Still trying to fetch through localhost.  Did I mess up my config, or is there a bug somewhere?

```json
$ cat .reactserverrc
{
  "port": "3000",
  "env": {
    "docker": {
      "host": "192.168.99.100"
    }
  }
}
```

**Comment by withinboredom** [comment]
I had issues with the config as well, I figured it was environmental ... guess I should open an issue. You can modify the script in package.json, here's mine for example:

```json
{
  "name": "reactFrontend",
  "version": "0.0.1",
  "description": "A react-server instance",
  "main": "HelloWorld.js",
  "scripts": {
    "start": "react-server-cli --host 192.168.99.100", 
    "test": "xo && nsp check"
  },
  "license": "Apache-2.0",
  "dependencies": {
    "babel-plugin-transform-runtime": "^6.8.0",
    "babel-preset-es2015": "^6.6.0",
    "babel-preset-react": "^6.5.0",
    "babel-runtime": "^6.6.1",
    "react": "~0.14.2",
    "react-dom": "~0.14.2",
    "react-server": "^0.2.10",
    "react-server-cli": "^0.2.10",
    "superagent": "1.2.0"
  },
  "devDependencies": {
    "eslint-config-xo-react": "^0.7.0",
    "eslint-plugin-react": "^5.1.1",
    "nsp": "^2.3.3",
    "xo": "^0.15.1"
  },
  "xo": {
    "esnext": true,
    "extends": "xo-react",
    "ignores": ["__clientTemp/**/*"]
  }
}
```

**Comment by withinboredom** [comment]
You can verify that the environment is set right by changing the hello world:

```javascript
Hello, World: {process.env.NODE_ENV}
```

**Comment by withinboredom** [comment]
I think I found the bug ... trying a fix and PR if that was it. Not really setup to debug CLI scripts, so console.logging all over ;)

**Comment by withinboredom** [comment]
Yep, fixed in #206 ... that was a fun one.

**Comment by gigabo** [comment]
> modify the script in package.json

Sweet, that worked.  Thanks.

Mind rebasing this branch to pick up your fix from #206 now that it's merged?

I'll see if I can track down what's going wrong with the tests.

**Comment by gigabo** [comment]
@doug-wade - Uhh.... why aren't the tests passing here?  Seems right.

**Comment by doug-wade** [comment]
@withinboredom The bug was in the tests, not in your changes 🐑 .  #207 should unblock you

**Comment by doug-wade** [comment]
@withinboredom I think its failing with a (well-obscured) linter failure; I've added a line comment

File: packages/generator-react-server/generators/app/index.js [hunk 8]
```diff
@@ -14,6 +14,11 @@ module.exports = yeoman.Base.extend({
 			name: 'name',
 			message: 'What would you like to call your app?',
 			default: this.appname
+		},{
```
Comment (doug-wade) [inline packages/generator-react-server/generators/app/index.js]: I think this is causing the linter failure that's failing your build; it should be `}, {`

Comment (gigabo) [inline packages/generator-react-server/generators/app/index.js]: @doug-wade - How did you track that down?  I don't see anything about it in travis...

Comment (doug-wade) [inline packages/generator-react-server/generators/app/index.js]: I ran the branch locally to find #207 and when I verified that it made the existing tests pass, I noticed there was still a linter failure.

**Comment by doug-wade** [comment]
lgtm 👍

**Comment by gigabo** [comment]
This is awesome!  Thanks @withinboredom!

I hope we'll see more cool stuff from you soon. 👹

**Comment by gigabo** [comment]
BTW - I'm testing out a changelog generator.

The next version's shaping up to be the @withinboredom show:

---

## Unreleased (2016-05-25)

#### Bug fix
* `react-server-cli`
  * [#206](https://github.com/redfin/react-server/pull/206) Default options for cli overwrite config file options. ([@withinboredom](https://github.com/withinboredom))

#### Enhancement
* `generator-react-server`
  * [#200](https://github.com/redfin/react-server/pull/200) Add docker to yeoman generator. ([@withinboredom](https://github.com/withinboredom))

#### Commiters: 1
- [withinboredom](https://github.com/withinboredom)

**Pull request merged**